### ISSUE TITLE
How to calculate mAP for each category for  waymo 2D object detection

### ISSUE THREAD

**Issue by gopikrishnabs** [shinya7y/UniverseNet#36]
I would like to know how to get mAP for each class according to waymo metrics. I tried seeing waymo's official quick start page but the ground truth is available for the entire validation set. I would like to test for only a subset of it, will the annotations file be okay for ground truth?

**Comment by shinya7y** [comment]
We can get AP for each category by adding `evaluation = dict(classwise=True)` to configs. Please note that the values are different from those calculated by the Waymo's official scripts because our code calculates AP based on COCO scripts.

I don't understand your question about subset from the short description.

**Comment by gopikrishnabs** [comment]
Do I need to create another data format (like waymo format or kitti format) to compute AP according to the official ways script, or would the repo *WaymotoCoco* be enough? 
I have trained on the training data for the waymo dataset now, I need to get the validation and test results so that I can know how the model is working(according to waymo AP computation) and what tweaks I need to make?. 
Since the data format is important for producing the bin file(for submission) I was wondering whether I can use the prepared data format.

In your research paper, you have validated on a subset of the *validation set i.e f0val* the image is below

![image](https://user-images.githubusercontent.com/53032421/169701507-8b195cc6-0841-4990-b2f9-b692e154fe4f.png)

But to compute AP according to waymo's official script it takes 2 files * 1 ground truth & 1 prediction file* to give results. 
The waymo data set bucket provides only the ground truth file for the entire validation set, but I need to validate on only a subset of the validation data for that is the annotation file which was created according to the "WaymotoCOCO" format can be used?

**Comment by shinya7y** [comment]
WaymoCOCO converts the **input data format** to train detectors. For evaluation with Waymo's official script, we need to convert the **output data format** of detectors.
By using [`--format-only`](https://github.com/shinya7y/UniverseNet/blob/9958e6d6b139584b5bf5f18d70a5deb514c3fb26/tools/test.py#L49-L54) option, [this code](https://github.com/shinya7y/UniverseNet/blob/9958e6d6b139584b5bf5f18d70a5deb514c3fb26/mmdet/datasets/waymo_open.py#L215-L240) converts results to a list of dicts, and it is saved to `f'{outfile_prefix}.bbox.pkl'`. You can read the pkl file in your script to produce a bin file.

> In your research paper, you have validated on a subset of the validation set i.e f0val the image is below

I compute the results in Table 10 using COCO script.

> But to compute AP according to waymo's official script it takes 2 files * 1 ground truth & 1 prediction file* to give results.
> The waymo data set bucket provides only the ground truth file for the entire validation set, but I need to validate on only a subset of the validation data for that is the annotation file which was created according to the "WaymotoCOCO" format can be used?

I haven't written scripts to compute AP on a subset using Waymo's official script. You can prepare [`waymococo_full/val2020`](https://github.com/shinya7y/WaymoCOCO#waymococo-full) to compute AP on the full val set, or write scripts by yourself to compute AP on a subset.

**Comment by gopikrishnabs** [comment]
I have got the .pkl file after running the testing script but it is of the format ( like a list where each index has 3 entries like ans[0][0],ans[0][1],ans[0][2]) 

(Is the achieved output is in the correct format)
1 entry of the result file is as below 

 [array([[7.53474731e+02, 6.16870605e+02, 8.42561707e+02, 6.89327820e+02,
        9.31835651e-01],
       [5.46413513e+02, 6.17624939e+02, 6.48374817e+02, 7.07772339e+02,
        9.08455610e-01],
       [6.52498840e+02, 6.24393616e+02, 7.77826233e+02, 7.15765442e+02,
        9.07803774e-01],
       [7.89793823e+02, 6.11643188e+02, 8.56424866e+02, 6.72305725e+02,
        8.45687687e-01],
       [1.03699902e+03, 6.02340210e+02, 1.09876416e+03, 6.43325928e+02,
        7.88908601e-01],
       [7.30879089e+02, 5.91809692e+02, 7.75247192e+02, 6.29956238e+02,
        7.18594372e-01],
       [1.02673193e+03, 5.92624329e+02, 1.07460706e+03, 6.36211304e+02,
        7.15026200e-01],
       [8.63896973e+02, 5.87437195e+02, 9.15828674e+02, 6.11548340e+02,
        6.74908161e-01],
       [6.36844971e+02, 6.05150696e+02, 7.12671082e+02, 6.54044922e+02,
        6.72331691e-01],
       [7.06556030e+02, 5.98389526e+02, 7.47868225e+02, 6.29081299e+02,
        6.47798240e-01],
       [1.00593939e+03, 5.86609314e+02, 1.03943994e+03, 6.23031250e+02,
        6.47352457e-01],
       [8.32942383e+02, 5.84528381e+02, 8.68671265e+02, 6.16419861e+02,
        6.44672632e-01],
       [6.30490112e+02, 6.21994324e+02, 6.71956299e+02, 6.79552490e+02,
        5.45769930e-01],
       [8.00477539e+02, 5.93927429e+02, 8.24221130e+02, 6.12919861e+02,
        5.00488281e-01],
       [9.49358215e+02, 5.80110718e+02, 9.74766113e+02, 5.98887512e+02,
        4.90235627e-01],
       [6.32666748e+02, 6.15592163e+02, 6.84963623e+02, 6.70234985e+02,
        4.57621485e-01],
       [8.75674988e+02, 5.81573486e+02, 9.18903259e+02, 6.04211609e+02,
        4.42154676e-01],
       [8.10368469e+02, 5.93499207e+02, 8.33486938e+02, 6.11535034e+02,
        4.36862260e-01],
       [7.90187439e+02, 5.96101196e+02, 8.15090332e+02, 6.15109619e+02,
        4.26799357e-01],
       [9.51028992e+02, 5.76092468e+02, 9.81471924e+02, 5.97849121e+02,
        4.14424807e-01],
       [9.79849304e+02, 5.83697998e+02, 9.99054443e+02, 6.06896790e+02,
        3.67497236e-01],
       [9.44374390e+02, 5.64824585e+02, 9.91891663e+02, 5.95302124e+02,
        3.66589725e-01],
       [7.39430542e+02, 5.91281250e+02, 7.83507263e+02, 6.24016846e+02,
        3.65683168e-01],
       [9.86245422e+02, 5.82909363e+02, 1.00704065e+03, 6.08688721e+02,
        3.46873075e-01],
       [7.80147827e+02, 5.94079407e+02, 8.03174683e+02, 6.16215820e+02,
        3.03215712e-01],
       [9.61736084e+02, 5.70799561e+02, 1.00113330e+03, 5.97855408e+02,
        3.01362038e-01],
       [1.01442047e+03, 5.94686951e+02, 1.03998669e+03, 6.29945679e+02,
        2.68365800e-01],
       [9.12641724e+02, 5.70360352e+02, 9.29913147e+02, 5.87236755e+02,
        2.65880466e-01],
       [9.80947388e+02, 5.73679565e+02, 1.00535345e+03, 6.00231750e+02,
        2.58887708e-01],
       [9.35385010e+02, 5.82363770e+02, 9.53579590e+02, 6.04686768e+02,
        2.52752751e-01],
       [8.29815308e+02, 5.91873169e+02, 8.64224976e+02, 6.21657349e+02,
        2.52199829e-01],
       [1.08788757e+03, 5.97270691e+02, 1.11419775e+03, 6.36174500e+02,
        2.49630138e-01],
       [1.07149719e+03, 5.92253418e+02, 1.09998181e+03, 6.25036987e+02,
        2.37929881e-01],
       [7.53925415e+02, 5.88527527e+02, 7.87090515e+02, 6.18175232e+02,
        2.36516222e-01],
       [8.17152771e+02, 6.13694519e+02, 8.59218994e+02, 6.71406006e+02,
        2.34056309e-01],
       [8.97475525e+02, 5.75190186e+02, 9.21120667e+02, 5.93425903e+02,
        2.32658640e-01],
       [1.09790210e+03, 5.94727966e+02, 1.12171704e+03, 6.45757874e+02,
        2.19672099e-01],
       [8.05029236e+02, 5.87059509e+02, 8.26846008e+02, 6.05926514e+02,
        2.13705182e-01],
       [9.30326721e+02, 5.82522705e+02, 9.47337891e+02, 6.08112305e+02,
        2.10603908e-01],
       [6.91268066e+02, 6.01067200e+02, 7.38545837e+02, 6.28590149e+02,
        2.10279375e-01],
       [1.09732129e+03, 5.90146118e+02, 1.11739600e+03, 6.28131531e+02,
        2.06253812e-01],
       [1.08682727e+03, 5.91243713e+02, 1.10811072e+03, 6.23710449e+02,
        2.04183161e-01],
       [9.49019165e+02, 5.61182068e+02, 9.99176270e+02, 5.87596802e+02,
        2.00871006e-01],
       [7.91041199e+02, 5.97862732e+02, 8.23221863e+02, 6.18513977e+02,
        1.89071491e-01],
       [1.03583887e+03, 5.85474426e+02, 1.07307593e+03, 6.20322205e+02,
        1.88921809e-01],
       [6.18768982e+02, 6.19020996e+02, 6.60173950e+02, 6.91658936e+02,
        1.82717010e-01],
       [8.02927673e+02, 6.00181213e+02, 8.49000061e+02, 6.34072754e+02,
        1.82134405e-01],
       [7.13113892e+02, 6.12837097e+02, 7.61376770e+02, 6.43480042e+02,
        1.77240521e-01],
       [6.64130798e+02, 6.04371460e+02, 7.23916626e+02, 6.40379639e+02,
        1.74269706e-01],
       [9.12986694e+02, 5.62915894e+02, 9.33144287e+02, 5.81947205e+02,
        1.72311082e-01],
       [9.13046997e+02, 5.79763367e+02, 9.31936157e+02, 6.01436523e+02,
        1.66404307e-01],
       [8.84436157e+02, 5.75353149e+02, 9.18631836e+02, 5.93945129e+02,
        1.65863156e-01],
       [5.39942932e+02, 6.20324524e+02, 5.62015564e+02, 6.78341675e+02,
        1.57914460e-01],
       [9.93278442e+02, 5.74374756e+02, 1.01971118e+03, 6.01179810e+02,
        1.56749219e-01],
       [9.51051697e+02, 5.84023254e+02, 9.69561890e+02, 6.04968506e+02,
        1.53042153e-01],
       [9.24369690e+02, 5.71518616e+02, 9.43682312e+02, 5.90399658e+02,
        1.50278062e-01],
       [9.34921326e+02, 5.59150757e+02, 9.88548645e+02, 5.87323792e+02,
        1.49283171e-01],
       [1.10698889e+03, 5.88835327e+02, 1.12952869e+03, 6.30938110e+02,
        1.41637221e-01]], dtype=float32), array([[8.22258118e+02, 6.69762634e+02, 1.02988586e+03, 1.02203082e+03,
        9.13058281e-01],
       [1.95888786e+01, 6.61287537e+02, 7.72427292e+01, 8.00755615e+02,
        8.72781694e-01],
       [1.38228064e+03, 5.89492981e+02, 1.41975818e+03, 6.85020935e+02,
        8.26711774e-01],
       [1.13169788e+03, 5.88534302e+02, 1.16373792e+03, 6.69570679e+02,
        7.52012551e-01],
       [1.10921484e+03, 5.95513916e+02, 1.13786841e+03, 6.68577087e+02,
        6.92641973e-01],
       [1.11758765e+03, 5.91825928e+02, 1.15415417e+03, 6.70167236e+02,
        6.33863628e-01],
       [1.15235913e+03, 5.91544739e+02, 1.17708740e+03, 6.67406189e+02,
        5.14644265e-01],
       [1.10173279e+03, 5.96021179e+02, 1.13000061e+03, 6.58740234e+02,
        4.02626961e-01],
       [3.34422485e+02, 6.40104431e+02, 3.68943634e+02, 7.27010925e+02,
        3.87226760e-01],
       [5.31019531e+02, 6.11972046e+02, 5.51587158e+02, 6.72597229e+02,
        3.66136342e-01],
       [9.30611694e+02, 5.83380310e+02, 9.46659180e+02, 6.08615906e+02,
        3.62969190e-01],
       [1.09790210e+03, 5.94727966e+02, 1.12171704e+03, 6.45757874e+02,
        3.49976927e-01],
       [4.16025726e+02, 6.14758545e+02, 4.39131866e+02, 6.69120667e+02,
        3.42021883e-01],
       [1.09732129e+03, 5.90146118e+02, 1.11739600e+03, 6.28131531e+02,
        2.81800717e-01],
       [1.07891882e+03, 5.81309326e+02, 1.09454907e+03, 6.05096558e+02,
        2.68941432e-01],
       [5.24770325e+02, 6.11284607e+02, 5.46794006e+02, 6.74782532e+02,
        2.67790943e-01],
       [9.34252502e+02, 5.83663208e+02, 9.50852112e+02, 6.07675659e+02,
        2.66261876e-01],
       [5.32150024e+02, 6.07864563e+02, 5.51785461e+02, 6.50228699e+02,
        2.46533319e-01],
       [1.08682727e+03, 5.91243713e+02, 1.10811072e+03, 6.23710449e+02,
        2.42205635e-01],
       [1.09971509e+03, 5.82372498e+02, 1.11730957e+03, 6.08645691e+02,
        2.40061253e-01],
       [5.39722412e+02, 6.15936096e+02, 5.60446350e+02, 6.69518311e+02,
        2.35811576e-01],
       [5.78090332e+02, 6.03754150e+02, 5.96205444e+02, 6.28891663e+02,
        2.34757334e-01],
       [5.73957825e+02, 6.04501404e+02, 5.91864319e+02, 6.31106689e+02,
        2.19672099e-01],
       [1.08514612e+03, 5.79554749e+02, 1.10077942e+03, 6.08025452e+02,
        2.08339944e-01],
       [3.39805298e+02, 6.41724854e+02, 3.74521332e+02, 7.05129822e+02,
        2.05615044e-01],
       [1.10698889e+03, 5.88835327e+02, 1.12952869e+03, 6.30938110e+02,
        2.04659641e-01],
       [3.33990784e+02, 6.39550110e+02, 3.64802673e+02, 6.97337341e+02,
        1.95901573e-01],
       [3.45851227e+02, 6.46039612e+02, 3.74599518e+02, 7.41310730e+02,
        1.95440486e-01],
       [9.79849304e+02, 5.83697998e+02, 9.99054443e+02, 6.06896790e+02,
        1.90874800e-01],
       [6.10130615e+02, 6.08372803e+02, 6.27528564e+02, 6.26729736e+02,
        1.76813692e-01],
       [5.33796753e+02, 6.19495483e+02, 5.54799561e+02, 6.81623352e+02,
        1.75538078e-01],
       [1.33977539e+03, 5.94351685e+02, 1.37615881e+03, 6.72683167e+02,
        1.71477020e-01],
       [5.69626038e+02, 6.07090027e+02, 5.90455933e+02, 6.37216553e+02,
        1.71061188e-01],
       [5.38382629e+02, 6.08285461e+02, 5.59376038e+02, 6.45827209e+02,
        1.62645862e-01],
       [1.08788757e+03, 5.97270691e+02, 1.11419775e+03, 6.36174500e+02,
        1.59741014e-01],
       [9.49358215e+02, 5.80110718e+02, 9.74766113e+02, 5.98887512e+02,
        1.58434615e-01],
       [4.65377274e+01, 6.63861511e+02, 8.41644287e+01, 7.76082520e+02,
        1.56878337e-01],
       [1.07149719e+03, 5.92253418e+02, 1.09998181e+03, 6.25036987e+02,
        1.52536526e-01],
       [9.86245422e+02, 5.82909363e+02, 1.00704065e+03, 6.08688721e+02,
        1.49283171e-01],
       [1.10600317e+03, 5.82089417e+02, 1.12471838e+03, 6.09460571e+02,
        1.48540571e-01],
       [2.79811798e+02, 6.36874207e+02, 3.05143799e+02, 7.25028381e+02,
        1.45843774e-01]], dtype=float32), array([[9.3061169e+02, 5.8338031e+02, 9.4665918e+02, 6.0861591e+02,
        1.6885695e-01]], dtype=float32)]

 how can I get the score for each category from this file?

I have gone through the sample_detection_prediction file on waymo. Is the conversion of the .pkl file to prediction file takes place like this? 

***context_id, frame_no, bbox..... , bbox height,width,... = each entry in .pkl file***?

2). Do I need to add any arguments for the ***--format-only*** argument I observed that inline https://github.com/shinya7y/UniverseNet/blob/9958e6d6b139584b5bf5f18d70a5deb514c3fb26/tools/test.py#L259 it calls for the method to get the data in a list of the dictionary format but the called method returns https://github.com/shinya7y/UniverseNet/blob/9958e6d6b139584b5bf5f18d70a5deb514c3fb26/mmdet/datasets/waymo_open.py#L431 but is it received & saved somewhere? I am not aware where the **context_id, and frame_no** are present in that .pkl file. kindly clarify

Thank you

**Comment by shinya7y** [comment]
2). `dict_results` will be saved (dumped) by [these lines](https://github.com/shinya7y/UniverseNet/blob/9958e6d6b139584b5bf5f18d70a5deb514c3fb26/mmdet/datasets/waymo_open.py#L323-L325).
We can specify `outfile_prefix` using an additional argument, e.g., `--eval-options "outfile_prefix=results_fulltest_epoch_7"`.
The usage of `outfile_prefix` for WOD is similar to [`jsonfile_prefix`](https://github.com/open-mmlab/mmdetection/blob/master/docs/en/1_exist_data_model.md) for COCO.

**Comment by gopikrishnabs** [comment]
Hello, I am now using the GCP AI notebook instance but to complete authentication the UI is not taking user response(The user input box is not working properly). I wonder how I can get authenticated to download Waymo dataset. In colab it's easy but in GCP I am facing this issue. Any commands or steps would be helpful.

**Comment by shinya7y** [comment]
I don't know more than https://github.com/shinya7y/WaymoCOCO#download

I will close this issue. Please feel free to open/reopen issues if you find **bugs** in this repo.
I cannot provide further support on how to use this repo or issues unrelated to this repo.
